crates/uv-build-backend/src/lib.rs

```diff
@@ -342,6 +351,126 @@ pub fn build(
+) -> Result<SourceDistFilename, Error> {
+    let contents = fs_err::read_to_string(source_tree.join("pyproject.toml"))?;
+    let pyproject_toml = PyProjectToml::parse(&contents)?;
+    pyproject_toml.check_build_system(uv_version);
+
+    let filename = SourceDistFilename {
+        name: pyproject_toml.name().clone(),
+        version: pyproject_toml.version().clone(),
+        extension: SourceDistExtension::TarGz,
+    };
+
+    let top_level = format!("{}-{}", pyproject_toml.name(), pyproject_toml.version());
+
+    let source_dist_path = source_dist_directory.join(filename.to_string());
+    let tar_gz = File::create(&source_dist_path)?;
+    let enc = GzEncoder::new(tar_gz, Compression::default());
+    let mut tar = tar::Builder::new(enc);
+
+    let metadata = pyproject_toml
+        .to_metadata(source_tree)?
+        .core_metadata_format();
+
+    let mut header = Header::new_gnu();
+    header.set_size(metadata.bytes().len() as u64);
+    header.set_mode(0o644);
+    header.set_cksum();
+    tar.append_data(
+        &mut header,
+        Path::new(&top_level).join("PKG-INFO"),
+        Cursor::new(metadata),
+    )
+    .map_err(|err| Error::TarWrite(source_dist_path.clone(), err))?;
+
+    let includes = ["src/**/*", "pyproject.toml"];
+    let mut include_builder = GlobSetBuilder::new();
+    for include in includes {
+        include_builder.add(Glob::new(include).expect("TODO"));
+    }
+    let include_matcher = include_builder.build().expect("TODO");
+
+    let excludes = ["__pycache__", "*.pyc", "*.pyo"];
+    let mut exclude_builder = GlobSetBuilder::new();
+    for exclude in excludes {
+        exclude_builder.add(Glob::new(exclude).expect("TODO"));
+    }
+    let exclude_matcher = exclude_builder.build().expect("TODO");
+
+    // TODO(konsti): Add files linked by pyproject.toml
+
+    for file in WalkDir::new(&source_tree).into_iter().filter_entry(|dir| {
+        let relative = dir
+            .path()
+            .strip_prefix(&source_tree)
+            .expect("walkdir starts with root");
+        // TODO(konsti): Also check that we're matching at least a prefix of an include matcher.
+        !exclude_matcher.is_match(relative)
+    }) {
+        let entry = file.map_err(|err| Error::WalkDir {
+            root: source_tree.to_path_buf(),
+            err,
+        })?;
+        let relative = entry
+            .path()
+            .strip_prefix(&source_tree)
+            .expect("walkdir starts with root");
+        if !include_matcher.is_match(relative) {
+            trace!("Excluding {}", relative.user_display());
+            continue;
+        }
+        debug!("Including {}", relative.user_display());
+
+        let metadata = fs_err::metadata(entry.path())?;
+        let mut header = Header::new_gnu();
+        #[cfg(unix)]
+        {
+            header.set_mode(std::os::unix::fs::MetadataExt::mode(&metadata));
+        }
+        #[cfg(not(unix))]
+        {
+            header.set_mode(0o644);
```

Comment:
I'm not even sure what this means on non-Unix?

Reply:
The problem we want to avoid is that we're writing a `.tar.gz` on windows, setting a default 000 mode and then fail to read the unpacked files on unix.

There are other odd implications, e.g., when you check out a git repo and build a source dist from it, it may be different when built on windows and when built on unix: On windows, we lose the permissions that may be stored in git (as ntfs does not support them), while on unix, we copy the permissions to the archive.

In the wild, 644 and 755 seem to be the most popular defaults: https://github.com/search?q=header.set_mode&type=code

Reply:
Yeah, and I imagine at least on Unix, a user's umask will smooth things out even when `644` is undesirable.

Reply:
[Here's what bsdtar / libarchive does on Windows](https://github.com/libarchive/libarchive/blob/d60f3ea1b3bfd9f1dcd4040de9fca53fbedd829a/libarchive/archive_read_disk_windows.c#L2035C9-L2035C36). My C is a bit _rusty_ (hehe 🙄), but basically:

1. Start with read bits for everyone, i.e. user, group and other: 444
2. If the Windows file/dir is not readonly, add write bits for everyone (with bitwise OR): 222 (resulting in 666)
3. If it is a directory, add execute bits for everyone: 111
4. If it is a regular file ending in *.bat, *.cmd, or *.exe, add execute bits for everyone: 111

I feel like there ought to be a portable Rust crate for creating tar files that performs reasonably similar heuristics if you don't have a strong desire to roll your own. 😊

This will work fine while running as a normal user (non-root), since tar by default applies the normal user's umask when extracting a tar archive. From the [manual page](https://www.gnu.org/software/tar/manual/html_node/Option-Summary.html):

> ‘--no-same-permissions’
> When extracting an archive, subtract the user’s umask from files from the permissions specified in the archive. This is the default behavior for ordinary users.

And:

> ‘--preserve-permissions’
> ‘--same-permissions’
> ‘-p’
> When tar is extracting an archive, it normally subtracts the users’ umask from the permissions specified in the archive and uses that number as the permissions to create the destination file. Specifying this option instructs tar that it should use the permissions directly from the archive.

If someone does extract a random tar file from the Internet as root, they should know to use `--no-same-permissions` and similar flags, in order not to shoot themselves in the foot.